File: typescript/src/transport.ts
/**
 * Dials the network connection a client talks over. It is the seam where a
 * WebSocket implementation plugs in: the built-in `NodeTransport` speaks
 * RFC 6455 on `node:net`, `WebSocketTransport` wraps the platform's global
 * `WebSocket`, and wrapping `ws`, a React Native socket or an in-memory pipe
 * for tests means implementing these two interfaces and nothing else.
 */
export interface Transport {
  dial(url: string, options: DialOptions): Promise<Connection>;
}

/**
 * What the client needs the transport to negotiate: the subprotocols its
 * protocols speak, and the headers that authenticate the request — a cookie or
 * a token, since an Action Cable server authorizes the upgrade request itself.
 *
 * `signal` bounds the dial.
 */
export interface DialOptions {
  subprotocols?: string[];
  headers?: HeaderEntries;
  signal?: AbortSignal;
}

/**
 * Headers as a transport receives them. A `Headers` is one, and so is a plain
 * list of pairs — which is what lets a test hand a transport a value a
 * `Headers` would refuse, and see it neutralized rather than passed on.
 */
export type HeaderEntries = Headers | Iterable<[string, string]>;

/**
 * Headers as a caller writes them. The same three shapes `new Headers()`
 * takes, named here because the global `HeadersInit` is DOM's and this package
 * builds for both.
 */
export type HeaderInit = Headers | Record<string, string> | Array<[string, string]>;

/** Bounds one read or write. */
export interface TransferOptions {
  signal?: AbortSignal;
}

/**
 * One live connection. `read` and `write` are each called one at a time, but
 * `close` may be called while either is outstanding and must interrupt it.
 */
export interface Connection {
  /** What the server negotiated, empty if it named none. */
  readonly subprotocol: string;

  /**
   * The next complete message. It rejects once the connection is unusable,
   * including when the signal is aborted.
   */
  read(options?: TransferOptions): Promise<string>;

  /** Sends one text message. */
  write(payload: string, options?: TransferOptions): Promise<void>;

  close(): Promise<void>;
}

/**
 * A connection that can say why it is hanging up. `close` sends a close frame
 * with 1000 Normal Closure; `closeWithStatus` sends one with the code and
 * reason given, for a caller with something to tell the server. Both built-in
 * transports' connections implement it.
 */
export interface StatusCloser extends Connection {
  closeWithStatus(code: number, reason?: string): Promise<void>;
}

export function isStatusCloser(connection: Connection): connection is StatusCloser {
  return typeof (connection as StatusCloser).closeWithStatus === "function";
}

File: rust/src/transport.rs
use std::sync::Arc;

use async_trait::async_trait;
use http::HeaderMap;

use crate::error::Error;

/// Dials the network connection a client talks over. It is the seam where a network handler
/// plugs in: the built-in `WebSocketTransport` speaks RFC 6455 over tokio-tungstenite, and
/// wrapping another WebSocket crate, or an in-memory pipe for tests, means implementing
/// this and [`Conn`] and nothing else.
#[async_trait]
pub trait Transport: Send + Sync {
    /// Opens one connection. `options` carries the subprotocols the client's protocols
    /// negotiate under and the headers that authorize the request.
    async fn dial(&self, url: &str, options: DialOptions) -> Result<Box<dyn Conn>, Error>;
}

/// A shared transport dials like the one it shares, so an application can keep one and hand
/// it to every client it builds.
#[async_trait]
impl<T: Transport + ?Sized> Transport for Arc<T> {
    async fn dial(&self, url: &str, options: DialOptions) -> Result<Box<dyn Conn>, Error> {
        (**self).dial(url, options).await
    }
}

/// What the client needs the transport to negotiate: the subprotocols its protocols speak,
/// most preferred first, and the headers that authenticate the request, a cookie or a
/// token, since an Action Cable server authorizes the upgrade request itself.
#[derive(Debug, Clone, Default)]
pub struct DialOptions {
    /// The subprotocols to offer, most preferred first, with
    /// [`SUBPROTOCOL_UNSUPPORTED`](crate::SUBPROTOCOL_UNSUPPORTED) last.
    pub subprotocols: Vec<String>,
    /// The headers the upgrade request carries and nothing else ambient.
    pub headers: HeaderMap,
}

/// One live connection. The client reads from one task and writes from another, one write
/// at a time, and may call [`close`](Conn::close) concurrently with either. Closing has to
/// interrupt a read in flight so the reader learns the connection is gone.
#[async_trait]
pub trait Conn: Send + Sync {
    /// What the server negotiated, `None` if it named nothing.
    fn subprotocol(&self) -> Option<&str>;

    /// The next complete message. Fails once the connection is unusable, including after
    /// `close`.
    async fn read(&self) -> Result<Vec<u8>, Error>;

    /// Sends one text message.
    async fn write(&self, payload: &[u8]) -> Result<(), Error>;

    /// Hangs up, with 1000 Normal Closure where the protocol has a code to send. Safe to
    /// call twice.
    async fn close(&self);

    /// Hangs up with a code and reason of the caller's choosing, for a caller with
    /// something to tell the server. Where the Go client asks whether a connection also
    /// implements `StatusCloser`, a Rust caller can always call this: the default forwards
    /// to [`close`](Conn::close), so a transport whose library can't say why hangs up all
    /// the same. The built-in transport sends the code and as much of the reason as a
    /// control frame has room for.
    async fn close_with_status(&self, code: u16, reason: &str) {
        let _ = (code, reason);
        self.close().await;
    }
}
